packages/aws-cdk/lib/api/hotswap/s3-bucket-deployments.ts

```diff
@@ -0,0 +1,125 @@
+import { ISDK } from '../aws-auth';
+import { ChangeHotswapImpact, ChangeHotswapResult, HotswapOperation, HotswappableChangeCandidate, establishResourcePhysicalName } from './common';
+import { EvaluateCloudFormationTemplate } from './evaluate-cloudformation-template';
+
+/**
+ * This means that the value is required to exist by CloudFormation's API (or our S3 Bucket Deployment Lambda)
+ * but the actual value specified is irrelevant
+ */
+export const REQUIRED_BY_CFN = 'required-to-be-present-by-cfn';
+
+export async function isHotswappableS3BucketDeploymentChange(
+  logicalId: string, change: HotswappableChangeCandidate, evaluateCfnTemplate: EvaluateCloudFormationTemplate,
+): Promise<ChangeHotswapResult> {
+  // In old-style synthesis, the policy used by the lambda to copy assets Ref's the assets directly,
+  // meaning that the changes made to the Policy are artifacts that can be safely ignored
+  if (change.newValue.Type === 'AWS::IAM::Policy') {
+    return changeIsForS3DeployCustomResourcePolicy(logicalId, change, evaluateCfnTemplate);
+  }
+
+  if (change.newValue.Type !== 'Custom::CDKBucketDeployment') {
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+
+  // note that this gives the ARN of the lambda, not the name. This is fine though, the invoke() sdk call will take either
+  const functionName = await establishResourcePhysicalName(logicalId, change.newValue.Properties?.ServiceToken, evaluateCfnTemplate);
+  if (!functionName) {
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+
+  const customResourceProperties = await evaluateCfnTemplate.evaluateCfnExpression({
+    ...change.newValue.Properties,
+    ServiceToken: undefined,
+  });
+
+  return new S3BucketDeploymentHotswapOperation(functionName, customResourceProperties);
+}
+
+class S3BucketDeploymentHotswapOperation implements HotswapOperation {
+  public readonly service = 'custom-s3-deployment';
+
+  constructor(private readonly functionName: string, private readonly customResourceProperties: any) {
+  }
+
+  public async apply(sdk: ISDK): Promise<any> {
+    // JSON.stringify() doesn't turn the actual objects to strings, but the lambda expects strings
+    const resourceProperties = stringifyObject(this.customResourceProperties);
+
+    return sdk.lambda().invoke({
+      FunctionName: this.functionName,
+      // Lambda refuses to take a direct JSON object and requires it to be stringify()'d
+      Payload: JSON.stringify({
+        RequestType: 'Update',
+        ResponseURL: REQUIRED_BY_CFN,
+        PhysicalResourceId: REQUIRED_BY_CFN,
+        StackId: REQUIRED_BY_CFN,
+        RequestId: REQUIRED_BY_CFN,
+        LogicalResourceId: REQUIRED_BY_CFN,
+        ResourceProperties: resourceProperties,
+      }),
+    }).promise();
+  }
+}
+
+async function changeIsForS3DeployCustomResourcePolicy(
+  logicalId: string, change: HotswappableChangeCandidate, evaluateCfnTemplate: EvaluateCloudFormationTemplate,
```

Comment:
Can you rename the parameter `logicalId` to something like `iamPolicyLogicalId`? I think it's a little confusing right now.